src/textinput/TextInput.ts

```diff
@@ -0,0 +1,154 @@
+import { WidgetBase } from '@dojo/widget-core/WidgetBase';
+import { ThemeableMixin, ThemeableProperties, theme } from '@dojo/widget-core/mixins/Themeable';
+import { v, w } from '@dojo/widget-core/d';
+import Label, { LabelOptions } from '../label/Label';
+import * as css from './styles/textinput.css';
+
+export type TextInputType = 'text' | 'email' | 'number' | 'password' | 'search' | 'tel' | 'url';
+
+/**
+ * @type TextInputProperties
+ *
+ * Properties that can be set on a TextInput component
+ *
+ * @property describedBy	ID of an element that provides more descriptive text
+ * @property disabled			Prevents the user from interacting with the form field
+ * @property formId				ID of a form element associated with the form field
+ * @property invalid			Indicates the value entered in the form field is invalid
+ * @property label				Label settings for form label text, position, and visibility
+ * @property maxLength		Maximum number of characters allowed in the input
+ * @property minLength		Minimum number of characters allowed in the input
+ * @property name					The form widget's name
+ * @property placeholder	Placeholder text
+ * @property readOnly			Allows or prevents user interaction
+ * @property required			Whether or not a value is required
+ * @property type					Input type, e.g. text, email, tel, etc.
+ * @property value				The current value
+ * @property onBlur				Called when the input loses focus
+ * @property onChange			Called when the node's 'change' event is fired
+ * @property onClick			Called when the input is clicked
+ * @property onFocus			Called when the input is focused
+ * @property onInput			Called when the 'input' event is fired
+ * @property onKeyDown		Called on the input's keydown event
+ * @property onKeyPress		Called on the input's keypress event
+ * @property onKeyUp			Called on the input's keyup event
+ * @property onMouseDown	Called on the input's mousedown event
+ * @property onMouseUp		Called on the input's mouseup event
+ * @property onTouchStart	Called on the input's touchstart event
+ * @property onTouchEnd		Called on the input's touchend event
+ * @property onTouchCancel	Called on the input's touchcancel event
+ */
+export interface TextInputProperties extends ThemeableProperties {
+	describedBy?: string;
+	disabled?: boolean;
+	formId?: string;
+	invalid?: boolean;
+	label?: string | LabelOptions;
+	maxLength?: number | string;
+	minLength?: number | string;
+	name?: string;
+	placeholder?: string;
+	readOnly?: boolean;
+	required?: boolean;
```

Comment:
@dylans do you think we need anything more than valid/invalid for validation? I'm not sure it's necessary to differentiate incomplete from invalid. I'm also wondering if I should change the property name to `valid`, just because it sounds better to me.